Support for test cases with multiple data input

MichaelHuth opened the pull request:
Should solve https://github.com/byte-physics/igor-unit-testing-framework/issues/5

This enabled multi data test cases.
It also works in combination with test cases that monitor background activity.

MichaelHuth commented:
@t-b second revision with multi data test cases taking only a single argument of the same type as the wave result of associated data generator

t-b commented:
@MichaelHuth FYI. Rebase against master to make CI pass. This is required due to gitlab runner retagging.

t-b commented:
I've dived into the code and looks from the bird eye logic good!

Some things I noted:
- GetUTFTDGenerator should be a generic function for arbitrary tags. The matching should also be much stricter using a regex ala

```
•splitstring/E="//*[[:space:]]*\\QUTF_TD_DATAGENERATOR\\E(.*)$" str, str1; print V_flag; print str1
  1
   func
```

- Please use/create constants for the wave type return values
- Adding lists like `return testCaseList + testCaseMDList` fails if the first list is empty as in `";"`.
- Can you move the change `string testCaseList = FunctionList("!*_IGNORE", ";", "KIND:18,NPARAMS:0,VALTYPE:1,WIN:" + procWin)` into a separate commit and adapt the documentation? This is a compatibilty relevant change and I don't want to hide it.
- The dimension label reading from the data generator wave seems to be missing.
- Have you ideas for getting the testcase count for JUNIT/TAP correct?

MichaelHuth commented:
> * Have you ideas for getting the testcase count for JUNIT/TAP correct?

For TAP I have to move the handling after the test cases into the do while loop.

For JUNIT, we have juProps.testCaseList that is at the moment only used to fill juProps.juTS.tests with the number of items in that list. So I can replace that with a testCaseCount field, that stores the number of test cases for this Suite.

To get the count I have to add a function to determine to complete number of test cases after

```
		testCaseList = getTestCasesMatch(procWin, testCase, enableRegExpTC, err)
		if(err & TC_LIST_EMPTY)
			continue
		endif
```
is called for each test suite. As the function had been verified before they went through the matching filter, I can rely on the DimSizes of the waves from the multi data test cases data generator functions.

MichaelHuth commented:
@t-b The test case numbers for JUNIT and tap should be correct now. I dropped the = sign from the generator line and introduced a generic GetFunctionTag function. I also addressed the other comments.

t-b commented:
`DisplayHelpTopic "Backslash in Regular Expressions"`

MichaelHuth commented:
@t-b with the last commit also multi data support for reentry functions is implemented. The reentry function must have the same function format as the origin multi data test case or w/o multi data.
Should we also call incrError() if the UTF encounters a multi data reentry function of the wrong format?

t-b commented:
@MichaelHuth Could you rebase against latest master so that CI passes?

t-b commented:
Review:

General:
- Looks already really promising!
- [ ] example 12 and 13 are not linked in the documentation
- [ ] please change the module names of the new procedure files in internal_dev, they currently coincide with the examples.
- [ ] example 13 has mixed indentation
- [ ] How about cutting down example 13 to a minimum example? If you like to keep the current example I'm happy having it as example14 as well.

- [ ] Whitespace issues
```
$ git diff master.. --check
procedures/unit-testing-basics.ipf:2187: trailing whitespace.
+
procedures/unit-testing-basics.ipf:2228: trailing whitespace.
+
procedures/unit-testing-basics.ipf:2337: trailing whitespace.
+
procedures/unit-testing-basics.ipf:2341: trailing whitespace.
+
procedures/unit-testing-basics.ipf:2344: trailing whitespace.
+
procedures/unit-testing-basics.ipf:2349: trailing whitespace.
+
```
- [ ] No need for semicolons
```
$ git grep ";$" *.ipf | grep -v "//"
unit-testing-basics.ipf:                SplitString/E=expr funcLine, funcTag;
```

t-b commented on docu/sphinx/source/advanced.rst:
```diff
@@ -250,6 +250,112 @@ the `_REENTRY` function.
+
+	   print "Reentered test case with argument ", var
+	   PASS()
+   End
```
Indentation in the example looks odd.

t-b commented:
> Should we also call incrError() if the UTF encounters a multi data reentry function of the wrong format?

Yes I think that this is a good idea. Same for wrong reentry format.

MichaelHuth commented on procedures/unit-testing-basics.ipf:
```diff
@@ -897,13 +972,122 @@ static Function TestCaseEnd(testCase, keepDataFolder)
+			tcCount += 1
+		else
+			dgenFuncName = GetFunctionTag(testCase, UTF_TD_GENERATOR, UTF_TD_GENERATOR_L)
+			dgenFuncName = GetFullFunctionName(err, dgenFuncName, procWin)
```
Yes, if the full function name can be retrieved is already checked before in CheckFunctionSignaturesTCMD(). Here it is just used to setup the globals before entering a multi data test case.

MichaelHuth commented on procedures/unit-testing-basics.ipf:
```diff
@@ -897,13 +972,122 @@ static Function TestCaseEnd(testCase, keepDataFolder)
+				continue
+			endif
+			WAVE wGenerator = fDgen()
+			if(!((wType1 == 1 && WaveType(wGenerator, 1) == wType1 && WaveType(wGenerator) & wType0) || (wType1 != 1 && WaveType(wGenerator, 1) == wType1)))
```
Yes, missed that location when introducing the constants.

MichaelHuth commented on procedures/unit-testing-basics.ipf:
```diff
@@ -901,7 +901,7 @@ End
 	string procWin
 
-	string fList = FunctionList("!*_IGNORE", ";", "KIND:18,NPARAMS:0,WIN:" + procWin)
+	string fList = FunctionList("!*_IGNORE", ";", "KIND:18,NPARAMS:0,VALTYPE:1,WIN:" + procWin)
```
Added that point

MichaelHuth commented on procedures/unit-testing-basics.ipf:
```diff
@@ -746,6 +799,28 @@ static Function CheckAbortCondition(abortCode)
 	endif
 End
 
+/// Check if function marked as Multi Data test case
```
Updated function comment

MichaelHuth commented on procedures/unit-testing-basics.ipf:
```diff
@@ -998,6 +1010,43 @@ static function GetTestCaseCount(testCaseList, procWin)
+/// Checks functions signature of each multi data test case candidate
+/// returns 1 if ok, 0 otherwise
+/// when 1 is returned the wave type variable contain the format
+static Function GetFunctionSignatureTCMD(testCase, wType0, wType1)
```
done

t-b commented:
The indentation in the documentation is not correct:

![grafik](https://user-images.githubusercontent.com/1185677/56289120-89c8d980-6120-11e9-8ec0-1f51932800b9.png)

And that code block is also missing in the output.

Please merge once that is resolved.

t-b commented:
@MichaelHuth Don't merge. Example11-background.ipf is now broken. Please fix that, thanks.

MichaelHuth commented:
I think this fixes it, it was introduced when converting the multi data decision logic to CallTestCase.

t-b commented:
@MichaelHuth Thanks. I'll trust your testing.
